src/models/StandaloneSimilarity.py

```diff
@@ -1,27 +1,169 @@
-"""
-import random
+import gc
+import itertools
+import math
+from multiprocessing.pool import Pool
 
-from cleaning_functions import pandas_vector_to_list
-from job_description_feature_extraction import get_tfidf_similarity, extract_relevant_documents
+import numpy as np
+import os
+from sklearn.feature_extraction.text import TfidfVectorizer
+from tqdm import tqdm
 
-"""
+from cleaning_functions import pandas_vector_to_list
+from job_description_feature_extraction import (tfidf_vectorize,
+                                                cosine_knn, calc_cosine_sim,
+                                                extract_top_k_documents)
 from models.BaseModel import BaseModel
 
+DATA_QUERIES_VECTOR_NPZ = '../preprocessing_data/queries_vector.npz'
+DATA_CORPUS_VECTOR_NPZ = '../preprocessing_data/corpus_vector.npz'
+DATA_SIMILIARITY_MATRIX = '../preprocessing_data/similiarity_matrix'
+DATA_Y_TRAIN = '../preprocessing_data/y_train'
+DATA_Y_TEST = '../preprocessing_data/y_test'
+
+FORCE_LOAD = True
+
 
 class StandaloneSimilarity(BaseModel):
+    def __init__(self, train_size=0.75, test_size=None, force_load_all=True):
+        # overwrite constructor so we can work on cached data
+        # without loading the entire data into memory
+        if not os.path.exists(DATA_QUERIES_VECTOR_NPZ) or FORCE_LOAD:
+            super().__init__(train_size=train_size, test_size=test_size,
+                             force_load_all=force_load_all)
+
     def predict(self):
-        """
+        if os.path.exists(DATA_QUERIES_VECTOR_NPZ) and not FORCE_LOAD:
+            print('{}: loading precomputed data'.format(self.__class__.__name__))
+            self.load_precomputed_data()
+        else:
+            self.precomputed_similarity()
+
+        batch_size = 100
+        batch_elements = math.ceil(self.queries_vector.shape[0] / batch_size)
+        batch_queue = np.array_split(self.queries_vector.A, batch_elements)
+        print("starting batch computation of Similarity and KNN calculation")
+
+        # # multiple versions of calculating the prediction, some faster, some use more mem
+
+        # prediction = self.multiprocessor_batch_calc(batch_queue)
+        prediction = self.batch_calculation(batch_queue)
+        # prediction = self.individual_calculation()
+        # prediction = self.cosine_knn_calc()
+        # prediction = self.custom_knn_calculation(prediction)
+
+        train_avg_salary = sum(self.y_train) / len(self.y_train)
+        cleaned_predictions = [x if str(x) != 'nan' else train_avg_salary for x in prediction]
```

Comment:
tbh i don't understand why other models don't have that problem, but it happens if you predict on the basis of other Nans (similar to having a 0 in a chain of multiplications)

better solution might be to clean input data better - should i give that a try?

if we cant make a proper prediction, we just fall back to the average salary

that being said i have only seen a single case of this in the entire data set - but the error calculation will fail if the predictions contain a nan value